### ISSUE TITLE
MPAS "bus error",   "shepherd terminated"

### ISSUE THREAD

**Issue by timhoar** [NCAR/DART#23]
I'm in trouble running filter now, with the updated model_mod and obs_seq.out (w/ modified obs error variance in some obs types).
It worked well with ens_size = 3, but now failed w/ ens_size = 100.
I checked with different obs types separately, and the filter seemed to be running fine w/ individual obs type.
But when I tried to run them all together in the full ensemble, the filter fails.
At first, I got a bus error - no idea what it is -, then I increased the number of node (select) from 16 to 36, and I got the error as below.
```
MPT ERROR: Rank 334(g:334) received signal SIGBUS(7).
        Process ID: 49730, Host: r5i2n9, Program: /glade/work/xxxx/DART/models/mpas_atm/work/filter
        MPT Version: HPE MPT 2.19  02/23/19 05:30:09

MPT: --------stack traceback-------

MPT: -----stack traceback ends-----
MPT: On host r5i2n9, Program /glade/work/xxxx/DART/models/mpas_atm/work/filter, Rank 334, Process 49730: Dumping core on signal SIGBUS(7) into directory /glade/scratch/xxxx/MPAS_DART/MPASg
MPT: shepherd terminated: r5i2n9.ib0.cheyenne.ucar.edu - job aborting
duration_secs = 143
```

Tim, I put "#PBS -k eod" as you suggested.
And let me attach input.nml, jfyi.

Just to be clear, before I submitted my last changes to the branch, it worked just fine in my test w/ 100 members. So I really think nothing was wrong with the updates. 
But then I changed obs_seq.out w/ different error values and now the filter doesn't work any more.
Can you find any clues from the error message shown above? I'm stuck...

**Comment by timhoar** [comment]
I am recapping the emails that have been exchanged on this:

1) It was failing in a way such that there was no core file - despite the MPT message saying it was dumping core.
2) compiling with debug options on and running with all the namelist options for debugging on
```
   output_forward_op_errors = .true.
   output_timestamps        = .true.
   trace_execution          = .true.
   silence                  = .false.
```
led to a message that reported a problem that
```
 lower pressure larger than upper pressure at cellid        1121
```
This was traced back to being able to identify the location of the observation that caused the message (which should not have been a fatal condition)

> Here I'd like to identify which obs caused such an unusual case for p_low < p_high (where low and high means model levels).
> Although the module that spit out the message does not have lat/lon arrays - everything was passed as a column at that point -, I could find the location based on the cell id as below.
> 
```
           cellid    lon(deg)   lat(deg)    terrain(m)
 (0)     71025  48.6419   58.7127    155.684
 (1)     70644  49.8716   57.7980    151.181
 (2)     1121    50.6445    58.8476   181.668
```
> 
> Here I could find two things: i) these three cells are neighbor cells, meaning that the cell 71025 is surrounded by the cells 70644 and 1121. So this error basically happened at one spot. 
> ii) this case occurred in the lowest two levels, so I suspect that it is related to surface obs (which would make sense to me).

I then suggested to use 'obs_seq_to_netcdf.f90'  and then specify a teeny-tiny region to geographically subset the obs. -or- you could convert 'all' of them and use 'link_obs.m' to plot them all ... and then use the matlab 'link' brush to identify the region and observations in that region. This would also tell you the observation index in the original file.

> Tim,
> 
> Thanks for your suggestion!
> I found one radiosonde data in that crashed location, and extracted that from obs_seq.out using obs_sequence_tool.
> Then I ran the filter again w/ the one sounding temperature obs at the location and could reproduce the failure with the same p_low < p_high message.
> And I figured that the obs was actually located below the lowest model level - you know, sounding data is reported in pressure, and at the lowest level, the obs pressure was higher than than the hydrostatic pressure computed from the lowest model level in this case. 
> So I updated model_mod to take care of such a case as there is no need to loop over the model levels going up any further when the obs pressure is already higher than the lowest model level.
> Now it seems to be working.
> 
> Next, I expanded obs_seq.out from a single obs to the entire sounding profile at the location.
> Then, the filter was done after the first obs (e.g. at the lowest pressure level), with the message in the log file as below.
> 
> (/glade/scratch/xxxx/MPAS_DART/MPASg/MPASg.1.o2812990)
``` 
>  PE 0: filter: No more obs to assimilate, exiting main loop
>  PE 0: filter: End of main filter assimilation loop, starting cleanup
```
> It just finished nicely, with all the output files. What the hack.
> I double-checked obs_seq.out to find a total of 35 obs in the profile, but somehow the filter decided to be done after assimilating the first obs. In the output obs_seq.final, all the 35 obs was listed but w/ dart_qc = 4 except for the very first one.
> So I'm confused. How can this happen?
> 
> You can see my latest input.nml, obs_seq.out, and the output (obs_seq.olderr.raobT1.final) in the same directory, if you want.
> 
> Thanks again. It is very helpful to talk to you, indeed.

So I was able to get a copy of the updated _model_mod.f90_ and the input data necessary to recreate the problem. I tested it with an ensemble size of 3 to get it running and the code worked perfectly. I suspect that section of the code is only exercised for certain ensemble members with the right pressure conditions. This may be why it works with 3 but not 100 members. I currently have no explanation as to why it worked on the feature branch before being merged onto the main branch.

**Comment by timhoar** [comment]
Ensemble member 54 (and only 54) looks like the forward operator is failing for all observations **_but_** the lowest one.  All the other ensemble members look fine. I have something to debug!

**Comment by timhoar** [comment]
Since the forward operator for the prior was failing and I knew what ensemble member was causing the problem, I decided to use _perfect_model_obs_ (pmo) to diagnose the problem. _pmo_ and _filter_ use the same forward operator. _pmo_ is also small enough that I can request a node and run it without submitting batch jobs.

Setting `output_forward_op_errors   = .true.` writes out a file for each task with a line for each failure. The observations get distributed across tasks. Consequently, each line contains the observation number and the error code from the forward operator.  In this case, it looks like:
```
0[204] r6i6n31:/<2>thoar/mpas > cat forward_op_errors[123]
           2         988
          20         988
           3         988
          21         988
           4         988
          22         988
```
For this observation type, the forward operator is _model_interpolate()_, so I scrolled through the file for "988"  which is in that same block of code:
```
   ! Check if pressure at lower level is larger than at upper level
   ! (shouldn't happen).
   if(any(pressure(i, :) > pressure(i-1, :))) then
      if (debug > 0 .and. do_output()) then
         write(*, *) 'lower pressure larger than upper pressure at cellid',  cellid
         do e=1, ens_size
            write(*, '(A,3I4,2F9.2)') 'ens#, level nums, p_low, p_high: ', e,i-1,i,pressure(i-1,e),pressure(i,e)
         enddo
      endif
      where(pressure(i, :) > pressure(i-1, :)) ier(:) = 988
   endif
```
so - now to the real debugging.

**Comment by timhoar** [comment]
The `find_pressure_bounds` routine checks if the hydrostatic pressure is greater at a level above compared to the level immediately below. While this is rare, it does happen. As implemented, the check was causing any observation above the second layer to be rejected. A fix has been proposed and is being tested.

**Comment by timhoar** [comment]
Still haven't gotten to the bus error problem.

I'm using a 50-member ensemble.
I can assimilate the single radiosonde profile on 16 nodes, no problem.
When I use the full observation sequence file - everything else is the same - the job dies with:
```
  routine: set_location
  message:  latitude (  -90.0000025044782      ) is not within range [-90,90]
```
I ran the obs_sequence_tool to try to remove all the observations 'outside' [-90,90] to no avail - 
the proglem persists. We write the observations in ASCII, it is clearly a roundoff problem.

**Comment by timhoar** [comment]
I have removed the (radiosonde) obs at the South Pole and can run cleanly for the remainder of the radiosonde obs (about 25k obs) in about 8 minutes on 32 nodes with full compiler debug options.

Using all observation types, it took about 76 minutes to assimilate almost 300k observations. PBS summary is reporting it took a total of 1.2 Tb of memory.

Rerunning with light optimization instead of debugging.

**Comment by timhoar** [comment]
Summary:
Intel compiler, 'Full' debug, 300k obs, 32 nodes@36 cpu:

- 4608 seconds
- memory 1477906392kb 1443267.961 Mb 1409.441 Gb 1.376 Tb
- vmem 5074554048kb 4955619.188 Mb 4839.472 Gb 4.726 Tb
- memory 1.2 Gb/cpu

"light" optimization (-O -assume buffered_io)

- 2022 seconds
- memory 1476315296.0 kb 1441714.156 Mb 1407.924 Gb 1.375 Tb
- virtual 2696282324.0 kb 2633088.207 Mb 2571.375 Gb 2.511 Tb
- memory 1.2 Gb/cpu

"fast" optimization (-Ofast -assume buffered_io)

- 888 seconds
- memory 1476356092.0 kb 1441753.996 Mb 1407.963 Gb 1.375 Tb
- virtual 2709893952.0 kb 2646380.812 Mb 2584.356 Gb 2.524 Tb
- memory 1.2 Gb/cpu

**Comment by timhoar** [comment]
So I've run a lot of tests lately with two cases: The smaller case I simply ran filter with 50 ensemble members (some of the results were reported in the previous comment - repeated here for completeness) first with 1 radiosonde obs, then the whole profile, then all 300k obs of all types.
There were debugging settings that conflate the timing ... but just for the 50 member case.
If the table simply says `DIED`, there was no hints of why the job died. The `MPT: shepherd ...` errors occurred very early.

```
MPAS   : nVertLevels = 55 ;
         nCells      = 208898 ;
         nEdges      = 626688 ;
         nVertices   = 417792 ;

   ens_size                 = 50
   stages_to_write          = 'preassim', 'output'
   num_output_state_members = 50
   num_output_obs_members   = 50
   output_members           = .true.
   output_mean              = .true.
   output_sd                = .true.
   write_all_stages_at_end  = .false.

e50_16n_36p_36mp DIED (MPT: shepherd 33s in) memory  109267120kb  106706.172 Mb  104.205 Gb 0.102 Tb # 0.2 Gb/task
e50_16n_36p_18mp DIED (wallclock 2 hours)    memory  266514864kb  260268.422 Mb  254.168 Gb 0.248 Tb # 0.9 Gb/task
e50_16n_36p_18mp WORKED (4868s) (300k, dbg)  memory  446390920kb  435928.633 Mb  425.712 Gb 0.416 Tb # 1.5 Gb/task
e50_16n_18p_18mp WORKED (4006s) (300k, dbg)  memory  446333072kb  435872.141 Mb  425.656 Gb 0.416 Tb # 1.5 Gb/task
e50_32n_36p_36mp WORKED (68s) (35 obs)       memory   39372960kb   38450.156 Mb   37.549 Gb 0.037 Tb # 0.5 Gb/task
e50_32n_36p_36mp WORKED (2022s) (300k, O2)   memory 1476315296kb 1441714.156 Mb 1407.924 Gb 1.375 Tb # 1.2 Gb/task
e50_32n_36p_36mp WORKED (4608s) (300k, dbg)  memory 1477906392kb 1443267.961 Mb 1409.441 Gb 1.376 Tb # 1.2 Gb/task
e50_32n_36p_36mp WORKED (888s)  (300k, fast) memory 1476356092kb 1441753.996 Mb 1407.963 Gb 1.375 Tb # 1.2 Gb/task
```

And then I ran single instance of a higher resolutions case through pmo, and then used a 3 member ensemble. I got smarter about spreading out the tasks across both sockets on the node (using the 'omplace' command --  Thanks Brian!) 

```
When running with 18 tasks, I am using omplace -c 0-8,36-44  (9-9)
When running with  9 tasks, I am using omplace -c 0-3,36-40  (4-5)

MPAS   : nVertLevels = 55 ;
         nCells      = 835586 ;
         nEdges      = 2506752 ;
         nVertices   = 1671168 ;
         1 observation

pmo_1n_36p_18mp DIED (14s)                 memory 36240884kb 35391.488 Mb 34.562 Gb 0.034 Tb # 1.9 Gb/task
pmo_1n_36p_18mp KILL (3m50s) (Bus error)   memory 45349504kb 44286.625 Mb 43.249 Gb 0.042 Tb # 2.4 Gb/task
pmo_1n_36p_18mp KILL (1m47s) (Bus error)   memory 45169360kb 44110.703 Mb 43.077 Gb 0.042 Tb # 2.4 Gb/task
pmo_1n_36p_9mp  WORKED  (20s)              memory 25741376kb 25138.062 Mb 24.549 Gb 0.024 Tb # 2.7 Gb/task
pmo_1n_36p_9mp  WORKED  (18s)              memory 26067192kb 25456.242 Mb 24.860 Gb 0.024 Tb # 2.8 Gb/task
pmo_2n_36p_18mp WORKED  (23s)              memory 43672820kb 42649.238 Mb 41.650 Gb 0.041 Tb # 1.2 Gb/task
pmo_4n_36p_18mp WORKED  (31s)              memory 44443364kb 43401.723 Mb 42.384 Gb 0.041 Tb # 0.6 Gb/task
pmo_4n_36p_36mp DIED (MPT shepherd, 7s in) memory 36240884kb 35391.488 Mb 34.562 Gb 0.034 Tb

   ens_size                 = 3
   num_output_state_members = 3
   num_output_obs_members   = 3
   stages_to_write          = 'preassim', 'output'
   output_members           = .true.
   output_mean              = .true.
   output_sd                = .true.

e3_3n_36p_18mp DIED (MPT: shepherd, 7s in) memory 472kb 0.461 Mb
e3_3n_36p_9mp  DIED (26s)                  memory  31433848kb  30697.117 Mb  29.978 Gb 0.029 Tb # 1.1 Gb/task
e3_4n_36p_18mp WORKED (2m59s)              memory 390044068kb 380902.410 Mb 371.975 Gb 0.363 Tb # 5.2 Gb/task
e3_4n_36p_36mp DIED (MPT: shepherd, 5s in) memory      2388kb
e3_4n_36p_9mp  DIED (MPT: shepherd 38s in) memory  46591772kb  45499.777 Mb  44.433 Gb 0.043 Tb # 0.6 Gb/task
e3_4n_36p_9mp  DIED (MPT: shepherd 34s in) memory  72965904kb  71255.766 Mb  69.586 Gb 0.068 Tb # 1.9 Gb/task
e3_6n_36p_18mp DIED (14s)                  memory  46911120kb  45811.641 Mb  44.738 Gb 0.044 Tb # 0.4 Gb/task
e3_6n_36p_36mp DIED (6s)                   memory        32kb
```

**Comment by syha** [comment]
Tim, thanks a lot for your testing on this high-res case!
As it's a bit hard for me to interprete your test results, could you summarize what they all mean? Or do you need more tests to make some guidelines or recommendations to users?  
I just don't know how to take advantage of all your hard work on these...

**Comment by syha** [comment]
Tim,

So, would you recommend the "fast" optimization for my full cycling
experiment?
I just checked my mkmf.template and found

FFLAGS  = $(INCS) -O2 -m64 -heap-arrays -assume buffered_io

How should I change it to make the filter run faster?
Is there any downside to go with the fast optimization?

Thanks,

Soyoung

**Comment by timhoar** [comment]
I have not evaluated the effect of the optimization on the effectiveness of
the assimilation.
Generally speaking, higher optimizations will change answers, but with an
algorthm like the EAKF, and the fact we're using MPI - which might not
reproduce across processor counts due to latency - the only way to know for
sure is to run a test with -O2 (usually safe) and -Ofast (these are Intel
options), and compare the results to make sure nothing unexpected happens.

Given that -Ofast was twice as fast as -O2, I'd hope that -Ofast still
gives good results.

If you run the test, please let me know.

Tim Hoar
Data Assimilation Research Section
National Center for Atmospheric Research
thoar@ucar.edu
303.497.1708

**Comment by hkershaw-brown** [comment]
There is an issue on the DART_development repo:
https://github.com/NCAR/DART_development/issues/119

**Comment by hkershaw-brown** [comment]
MPAS silent death (no message, filter just dies) kills the dubgger. 

<img width="750" alt="Screen Shot 2020-11-18 at 4 40 29 PM" src="https://user-images.githubusercontent.com/20047007/99601852-83b27300-29bd-11eb-809c-4f48e432068e.png">

I'm not sure if silent death is typical for a Cheyenne job running out of memory or I should be expected a message from the OOM killer.

**Comment by syha** [comment]
That's scary.... But yes, it just died in my high-resolution tests (which seemed to be the out-of-memory issue).
We may need to check with cislhelp on this silent death...
